src/proxy/buffer.rs

```diff
@@ -1,91 +1,138 @@
 extern crate linkerd2_router as rt;
 
-use std::{fmt, marker::PhantomData};
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex, Weak};
+use std::time::Instant;
+use std::{error, fmt};
 
-use futures::{Future, Poll};
-use tower::buffer::Buffer;
+use futures::{Async, Future, Poll};
+use tokio::timer::Delay;
+use tower::buffer;
 
 use logging;
 use proxy::Error;
 use svc;
 
+pub trait ExtractDeadline<Req>: Clone {
+    fn extract(&self, req: &Req) -> Option<Instant>;
```

Comment:
I'm not sure how I feel about this name --- I think if I saw it elsewhere I wouldn't be clear what `extract` means? Maybe `fn extract_deadline`? Or, rename the trait to `HasDeadline` and just make this `fn deadline`?

Reply:
i'm -1 on a raw fn -- the signature gets gnarly with lifetimes

i'm +0 on Has -- We tend to use `Has..` with targets

`Extract` is what tower-web et al use to refer to pulling data out of requests, so it seemed relevant.

I do tink the fn should probably be called `deadline`...

Reply:
settled on just Deadline

Reply:
> i'm -1 on a raw fn -- the signature gets gnarly with lifetimes

I didn't mean a raw fn, I just meant renaming the trait fn to `extract_deadline`. 

`deadline` seems fine :+1: